plugin/trino-hive/src/main/java/io/trino/plugin/hive/avro/AvroHiveFileWriter.java

```diff
@@ -71,19 +74,24 @@ public AvroHiveFileWriter(
         verify(requireNonNull(fileSchema, "fileSchema is null").getType() == Schema.Type.RECORD, "file schema must be record schema");
         verify(inputColumnNames.size() == inputColumnTypes.size(), "column names must be equal to column types");
         // file writer will reorder input columns to schema, we just need to impute nulls for schema fields without input columns
-        ImmutableList.Builder<String> outputColumnNames = ImmutableList.<String>builder().addAll(inputColumnNames);
+        ImmutableList.Builder<String> outputColumnNames = ImmutableList.<String>builder();
         ImmutableList.Builder<Type> outputColumnTypes = ImmutableList.<Type>builder().addAll(inputColumnTypes);
-        Map<String, Schema> fields = fileSchema.getFields().stream().collect(Collectors.toMap(Schema.Field::name, Schema.Field::schema));
+        Map<String, String> lowerToGivenName = getLowerCaseToGivenFieldName(fileSchema);
+        Map<String, Field> fields = fileSchema.getFields().stream().collect(Collectors.toMap(Field::name, Function.identity()));
         for (String inputColumnName : inputColumnNames) {
-            Schema fieldSchema = fields.remove(inputColumnName);
-            if (fieldSchema == null) {
-                throw new AvroTypeException("File schema doesn't have input field " + inputColumnName);
+            Field field = fields.remove(inputColumnName);
+            if (field == null) {
+                if (!lowerToGivenName.containsKey(inputColumnName)) {
+                    throw new AvroTypeException("File schema doesn't have input field " + inputColumnName);
+                }
+                field = fields.remove(lowerToGivenName.get(inputColumnName));
```

Comment:
You could just always go through this map path. If the name is already lowercased you'll just have an identity map.